demo-app/src/main/java/cash/z/ecc/android/sdk/demoapp/util/BenchmarkTrace.kt

```diff
@@ -0,0 +1,106 @@
+package cash.z.ecc.android.sdk.demoapp.util
+
+import android.os.Looper
+import androidx.tracing.Trace
+import cash.z.ecc.android.sdk.ext.BenchmarkingExt
+import cash.z.ecc.android.sdk.internal.twig
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+interface BenchmarkTrace {
+    fun checkMainThread() {
+        check(Looper.getMainLooper().thread === Thread.currentThread()) {
+            "Should be called from the main thread, not ${Thread.currentThread()}."
+        }
+    }
+
+    val writeEventMutex: Mutex
+        get() = Mutex()
+
+    interface Event
+}
+
+object SyncBlockchainBenchmarkTrace : BenchmarkTrace {
+    private const val SECTION_BALANCE_SCREEN = "BALANCE_SCREEN" // NON-NLS
+    private const val SECTION_BLOCKCHAIN_SYNC = "BLOCKCHAIN_SYNC"   // NON-NLS
+    private const val SECTION_DOWNLOAD = "DOWNLOAD" // NON-NLS
+    private const val SECTION_VALIDATION = "VALIDATION" // NON-NLS
+    private const val SECTION_SCAN = "SCAN" // NON-NLS
+
+    @Suppress("MagicNumber")
+    suspend fun writeEvent(event: BenchmarkTrace.Event?) {
+        twig("New SyncBlockchain event: $event arrived.")
+        if (!BenchmarkingExt.isBenchmarking()) {
+            return
+        }
+        writeEventMutex.withLock {
+            checkMainThread()
+            when (event) {
+                Event.BALANCE_SCREEN_START -> Trace.beginAsyncSection(SECTION_BALANCE_SCREEN, 100)
+                Event.BALANCE_SCREEN_END -> Trace.endAsyncSection(SECTION_BALANCE_SCREEN, 100)
+                Event.BLOCKCHAIN_SYNC_START -> Trace.beginAsyncSection(SECTION_BLOCKCHAIN_SYNC, 200)
+                Event.BLOCKCHAIN_SYNC_END -> Trace.endAsyncSection(SECTION_BLOCKCHAIN_SYNC, 200)
+                Event.DOWNLOAD_START -> Trace.beginAsyncSection(SECTION_DOWNLOAD, 300)
+                Event.DOWNLOAD_END -> Trace.endAsyncSection(SECTION_DOWNLOAD, 300)
+                Event.VALIDATION_START -> Trace.beginAsyncSection(SECTION_VALIDATION, 400)
+                Event.VALIDATION_END -> Trace.endAsyncSection(SECTION_VALIDATION, 400)
+                Event.SCAN_START -> Trace.beginAsyncSection(SECTION_SCAN, 500)
+                Event.SCAN_END -> Trace.endAsyncSection(SECTION_SCAN, 500)
+                else -> { /* nothing to write */ }
+            }
+        }
+    }
+
+    enum class Event : BenchmarkTrace.Event {
```

Comment:
Could this be made a rich enum, where each value has a string property?  I think that might simplify the code and eliminate needing the constant strings declared separately.

Reply:
I had a solution with sealed class, but it was not helpful in manner of code-readability. Rich enums are the right way to go.